[FLINK-37515] Basic support for Blue/Green deployments

schongloo opened the pull request:
<!--
*Thank you very much for contributing to the Apache Flink Kubernetes Operator - we are happy that you want to help us improve the project. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix][docs] Fix typo in event time introduction` or `[hotfix][javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can read more on how we use GitHub Actions for CI [here](https://nightlies.apache.org/flink/flink-kubernetes-operator-docs-main/docs/development/guide/#cicd).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

This pull request adds basic support for Blue/Green deployments as outlined by FLIP-503 (https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=337677648).


## Brief change log
  - A new `FlinkBlueGreenDeploymentController` controller, with the capability of managing the lifecycle of these deployments.
  - A new corresponding CRD is introduced
  - The new `FlinkBlueGreenDeploymentController` will manage this CRD and hide from the user the details of the actual Blue/Green (Active/StandBy) jobs.
  - Delegate the lifecycle of the actual Jobs to the existing `FlinkDeployment` controller.

## Verifying this change

  - Test to verify a single basic deployment
  - Test to verify a proper transition Blue -> Green
  - Test to verify correct behavior when encountering an error before a transition
  - Test to verify correct behavior when encountering an error during a transition

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changes to the `CustomResourceDescriptors`: no
  - Core observer or reconciler logic that is regularly executed: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? docs / JavaDocs. UML diagrams for all scenarios added under `docs/content/docs`

### Overview Diagram
<img width="3853" height="3182" alt="image" src="https://github.com/user-attachments/assets/9a02450a-ef88-4032-a301-0a7032908851" />

davidradl commented on flink-kubernetes-operator-api/src/main/java/org/apache/flink/kubernetes/operator/api/bluegreen/DeploymentType.java:
```diff
@@ -0,0 +1,32 @@
+
+/** Enumeration of the two possible Flink Blue/Green deployment types. */
+public enum DeploymentType {
+    BLUE,
```
nit: could we have a description in the enum value please. It will be more self documenting then.

davidradl commented on flink-kubernetes-operator-api/src/main/java/org/apache/flink/kubernetes/operator/api/status/FlinkBlueGreenDeploymentState.java:
```diff
@@ -0,0 +1,26 @@
+
+/** Enumeration of the possible states of the blue/green transition. */
+public enum FlinkBlueGreenDeploymentState {
+    ACTIVE_BLUE,
```
could we have a descriptions for these enums as well please

davidradl replied:
I was thinking that the enum would have the value and the description - this could help if we need to access the description programatically, e.g. in a UI.

davidradl commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkBlueGreenDeploymentController.java:
```diff
@@ -0,0 +1,609 @@
+            // This phase requires rescheduling the reconciliation because the pod initialization
+            // could get stuck
+            //  (e.g. waiting for resources)
+            // TODO: figure out the course of action for error/failure cases
```
It seems risky to put in a capability without error failure cases being dealt with, is there a way to put this under a feature flag , until it is mature enough and handles errors and failures?

schongloo replied:
Thanks for pointing this out, this comment is outdated and I'll clean it up. The edge cases are addressed with the current set of tests.

davidradl commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkBlueGreenDeploymentController.java:
```diff
@@ -0,0 +1,609 @@
+                        .collect(Collectors.toList());
+
+        if (!badEvents.isEmpty()) {
+            LOG.warn("Bad events detected: " + badEvents);
```
nit: does this log entry format reasonably?

schongloo replied:
Yes, the events are logged like so:
```[Event(action=null, apiVersion=v1... message=MountVolume.SetUp failed for volume "checkpoint-volume" : hostPath type check failed: /tmp/checkpoints is not a directory, metadata=ObjectMeta(annotations={}, creationTimestamp=...```

davidradl commented:
I am concerned about the number of TODOs. Things like dealing with errors and timing conditions should be in place in the initial drop of code - unless we can mark this capability as beta or the like. WDYT @gyfora ?

schongloo commented:
@davidradl thanks for your review/comments. I've cleaned up all the outdated/invalid TODOs which pointed me 1 missing unit test, added. At this point the few remaining TODOs do not interfere with the main functionality and are not critical/open for discussion.

hansh0801 commented:
when this feature gonna be released?

schongloo commented:
Hi @hansh0801, trying to release it ASAP but I also want to make sure the logic and contracts are as robust as possible to minimize changes later. Looks pretty stable in my current testing and I've reopened it for review. Thanks!

hansh0801 commented:
cool ! and I'm also waiting for blue green phase 2 :)
https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=337677650

schongloo commented:
@hansh0801 indeed... Phase 2 will take a little bit longer due to its complexity but already working on it.

gyfora commented on flink-kubernetes-operator-api/src/main/java/org/apache/flink/kubernetes/operator/api/bluegreen/DeploymentType.java:
```diff
@@ -0,0 +1,38 @@
+
+    public static DeploymentType fromDeployment(FlinkDeployment flinkDeployment) {
+        String typeAnnotation =
+                flinkDeployment.getMetadata().getLabels().get(DeploymentType.class.getSimpleName());
```
Instead of using the label `DeploymentType` we should probably use a String constant that is less likely to clash with user annotations, such as:
`flink/blue-green-deployment-type`

gyfora commented on flink-kubernetes-operator-api/src/main/java/org/apache/flink/kubernetes/operator/api/bluegreen/TransitionMode.java:
```diff
@@ -0,0 +1,33 @@
+package org.apache.flink.kubernetes.operator.api.bluegreen;
+
+/** Possible transition modes supported by the `FlinkBlueGreenDeploymentController`. */
+public enum TransitionMode {
```
I would remove this from this PR and only introduce it when it's implemented as part of the next flip

gyfora commented on flink-kubernetes-operator-api/src/main/java/org/apache/flink/kubernetes/operator/api/spec/FlinkDeploymentTemplateSpec.java:
```diff
@@ -0,0 +1,60 @@
+    @JsonProperty("metadata")
+    private ObjectMeta metadata;
+
+    @JsonProperty("deploymentDeletionDelayMs")
+    private int deploymentDeletionDelayMs;
+
+    @JsonProperty("abortGracePeriodMs")
+    private int abortGracePeriodMs;
+
+    @JsonProperty("reconciliationReschedulingIntervalMs")
+    private int reconciliationReschedulingIntervalMs;
```
My gut feeling is that we will have more configurations in the past so maybe we should simply introduce a 


`Map<String, String> configuration` object and use Flink ConfigOptions for the timeout/interval settings. This is what we do basically everywhere else. And restrict the spec to key top level configs.

gyfora commented on flink-kubernetes-operator-api/src/main/java/org/apache/flink/kubernetes/operator/api/status/FlinkBlueGreenDeploymentStatus.java:
```diff
@@ -0,0 +1,58 @@
+    private String lastReconciledSpec;
+
+    /** Timestamp of last reconciliation. */
+    private Long lastReconciledTimestamp;
+
+    /** Computed from abortGracePeriodMs, timestamp after which the deployment should be aborted. */
+    private long abortTimestamp;
+
+    /** Timestamp when the deployment became READY/STABLE. Used to determine when to delete it. */
+    private long deploymentReadyTimestamp;
```
All timestamps should be String timestamps (Instant.toString)

gyfora commented on flink-kubernetes-operator-api/src/main/java/org/apache/flink/kubernetes/operator/api/status/FlinkBlueGreenDeploymentStatus.java:
```diff
@@ -0,0 +1,58 @@
+    private long deploymentReadyTimestamp;
+
+    /** Information about the TaskManagers for the scale subresource. */
+    private TaskManagerInfo taskManager;
```
Do we want to support this? I feel that the scale logic using TM count is not really used feature for regular FlinkDeployments either

gyfora replied:
We could remove it for now and only add it if there is a need

gyfora commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkBlueGreenDeploymentController.java:
```diff
@@ -0,0 +1,703 @@
+    }
+
+    public static void logAndThrow(String message) {
+        LOG.error(message);
```
Do we need to log this here? Maybe better to simply throw it and let error handling logic log it if needed

gyfora commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkBlueGreenDeploymentController.java:
```diff
@@ -0,0 +1,703 @@
+        }
+    }
+
+    private ObjectMeta getDependentObjectMeta(FlinkBlueGreenDeployment bgDeployment) {
+        ObjectMeta bgMeta = bgDeployment.getMetadata();
+        ObjectMeta objectMeta = new ObjectMeta();
+        objectMeta.setNamespace(bgMeta.getNamespace());
+        objectMeta.setOwnerReferences(
+                List.of(
+                        new OwnerReference(
+                                bgDeployment.getApiVersion(),
+                                true,
+                                false,
+                                bgDeployment.getKind(),
+                                bgMeta.getName(),
+                                bgMeta.getUid())));
+        return objectMeta;
+    }
```
Should be moved to a static utility class

gyfora commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkBlueGreenDeploymentController.java:
```diff
@@ -0,0 +1,703 @@
+        return objectMeta;
+    }
+
+    private static <T> T adjustNameReferences(
+            T spec,
+            String deploymentName,
+            String childDeploymentName,
+            String wrapperKey,
+            Class<T> valueType)
+            throws JsonProcessingException {
+        String serializedSpec = SpecUtils.serializeObject(spec, wrapperKey);
+        String replacedSerializedSpec = serializedSpec.replace(deploymentName, childDeploymentName);
+        return SpecUtils.deserializeObject(replacedSerializedSpec, wrapperKey, valueType);
+    }
```
Should be moved to a utility class

gyfora commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkBlueGreenDeploymentController.java:
```diff
@@ -0,0 +1,703 @@
+        deploymentStatus.setLastReconciledTimestamp(System.currentTimeMillis());
+    }
+
+    public void logPotentialWarnings(
+            FlinkDeployment flinkDeployment,
+            Context<FlinkBlueGreenDeployment> josdkContext,
+            long lastReconciliationTimestamp) {
```
We are already logging these errors as part of the FlinkDeployment logic, this is duplicated code/logic here that should be removed

gyfora commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkBlueGreenDeploymentController.java:
```diff
@@ -0,0 +1,703 @@
+                .rescheduleAfter(getReconciliationReschedInterval(bgDeployment));
+    }
+
+    private boolean isDeploymentReady(
+            FlinkDeployment deployment,
+            Context<FlinkBlueGreenDeployment> josdkContext,
+            FlinkBlueGreenDeploymentStatus deploymentStatus) {
```
We should not check for pods here, we can eliminate all this logic I believe. This is too low level on this controller and if we feel that some info is missing from the FlinkDeployment status then it should be added there

gyfora commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkBlueGreenDeploymentController.java:
```diff
@@ -0,0 +1,703 @@
+        return false;
+    }
+
+    private static Stream<PodResource> getDeploymentPods(
+            Context<FlinkBlueGreenDeployment> josdkContext, FlinkDeployment deployment) {
+        var namespace = deployment.getMetadata().getNamespace();
+        var deploymentName = deployment.getMetadata().getName();
+
+        return josdkContext
+                .getClient()
+                .pods()
+                .inNamespace(namespace)
+                .withLabel("app", deploymentName)
+                .resources();
+    }
```
probably not needed

gyfora commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkBlueGreenDeploymentController.java:
```diff
@@ -0,0 +1,703 @@
+                .resources();
+    }
+
+    private boolean hasSpecChanged(
+            FlinkBlueGreenDeploymentSpec newSpec, FlinkBlueGreenDeploymentStatus deploymentStatus) {
+
+        String lastReconciledSpec = deploymentStatus.getLastReconciledSpec();
+        String newSpecSerialized = SpecUtils.serializeObject(newSpec, "spec");
+
+        // TODO: in FLIP-504 check here the TransitionMode has not been changed
+
+        return !lastReconciledSpec.equals(newSpecSerialized);
```
We may want to use the DiffBuilder/ existing utilities to get the type of the spec change here and make sure we don't trigger deployments on non-upgrade changes. 

Things like operator config changes etc should not go through the Blue/Green flow I believe but simply applied on the currently active deployment.

gyfora commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkBlueGreenDeploymentController.java:
```diff
@@ -0,0 +1,703 @@
+                        .withName(deploymentName)
+                        .delete();
+
+        boolean deleted =
+                deletedStatus.size() == 1
+                        && deletedStatus.get(0).getKind().equals("FlinkDeployment");
+
```
do we want to wait here until the deletion is actually finished (the object is not there anymore?) that would require some additional logic here

davidradl commented on flink-kubernetes-operator-api/src/main/java/org/apache/flink/kubernetes/operator/api/status/FlinkBlueGreenDeploymentState.java:
```diff
@@ -0,0 +1,40 @@
+    TRANSITIONING_TO_BLUE,
+
+    /** Identifies the system is transitioning from "Blue" to "Green". */
+    TRANSITIONING_TO_GREEN,
```
what state are we in during shutdown?

davidradl commented on flink-kubernetes-operator-api/src/main/java/org/apache/flink/kubernetes/operator/api/utils/SpecUtils.java:
```diff
@@ -99,6 +99,28 @@ public static String writeSpecWithMeta(
         }
     }
 
+    public static String serializeObject(Object object, String wrapperKey) {
```
I see that `serializeObject` seems to be only called for specs and I see the previous method is called `writeSpecWithMeta`. I suggest being consistent with the method names, maybe `writeSpecAsJSON`.
The key seems to always be called with "spec", could we hard code it here - if this is the case.

schongloo replied:
Yes I like that name suggestion, thanks. And indeed I'm using it on `FlinkBlueGreenDeploymentSpec` objects only but, unlike other specs, it doesn't extend `AbstractFlinkSpec`.

davidradl commented on helm/flink-kubernetes-operator/templates/rbac.yaml:
```diff
@@ -101,6 +103,7 @@ rules:
       - flink.apache.org
     resources:
+      - flinkbluegreendeployments/status
       - flinkdeployments/status
```
this maybe a basic question. Are flinkbluegreendeployments a type of flinkdeployments. Would we expect to see them as flinkdeployments?

schongloo replied:
That's the original idea but we shifted to a `has-is` relationship rather than a `is-a` one.

maheepm-lyft commented:
Hi @schongloo @davidradl, I hope you're both doing well. I've been following this PR as blue-green deployments would solve a critical use case for our production needs. I noticed there have been no updates to this PR for a few weeks. Was wondering if there is anything blocking the release of this and if there is something myself or the OSS community could help with (testing, fixing any changes, etc). Thanks for all the work so far! 🙏

gyfora commented:
@schongloo can you please re-open this PR against the `main` branch?

schongloo commented:
> @schongloo can you please re-open this PR against the `main` branch?

Done @gyfora !

schongloo commented on flink-kubernetes-operator-api/src/main/java/org/apache/flink/kubernetes/operator/api/spec/FlinkBlueGreenDeploymentConfigOptions.java:
```diff
@@ -0,0 +1,54 @@
+    public static final int MIN_ABORT_GRACE_PERIOD_MS = 120000; // 2 mins
+
+    public static ConfigOptions.OptionBuilder operatorConfig(String key) {
+        return ConfigOptions.key(BLUE_GREEN_CONF_PREFIX + key);
```
I cannot add a direct reference to `KubernetesOperatorConfigOptions.` due to circular dependency, but I'll add the "kubernetes.operator." prefix accordingly

gyfora commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkBlueGreenDeploymentController.java:
```diff
@@ -0,0 +1,574 @@
+@ControllerConfiguration
+public class FlinkBlueGreenDeploymentController implements Reconciler<FlinkBlueGreenDeployment> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(FlinkDeploymentController.class);
```
Incorrect logger factory

gyfora commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkBlueGreenDeploymentController.java:
```diff
@@ -0,0 +1,574 @@
+                    bgDeployment, deploymentStatus, josdkContext, currentDeployment, nextState);
+        } else {
+            long delay = deletionTs - System.currentTimeMillis();
+            LOG.info("Rescheduling reconciliation (to delete) in {} ms.", delay);
```
probably better to log in seconds

gyfora commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkBlueGreenDeploymentController.java:
```diff
@@ -0,0 +1,574 @@
+
+        if (deploymentReadyTimestamp == 0) {
+            LOG.info(
+                    "Deployment marked ready on {}, rescheduling reconciliation in {} ms.",
```
Instead if `Deployment` we should write `FlinkDeployment` in all logs to avoid confusion. Also we should show seconds instead of milliseconds.

The timestamp is already in the log (when it was logged), not necessary to add

gyfora commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkBlueGreenDeploymentController.java:
```diff
@@ -0,0 +1,574 @@
+        } else {
+            // RETRY
+            var delay = abortTimestamp - System.currentTimeMillis();
+            LOG.info("Deployment '{}' not ready yet, retrying in {} ms", deploymentName, delay);
```
*FlinkDeployment, log reschedule in seconds.

gyfora commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkBlueGreenDeploymentController.java:
```diff
@@ -0,0 +1,574 @@
+                        && deletedStatus.get(0).getKind().equals("FlinkDeployment");
+
+        if (!deleted) {
+            LOG.info("Deployment '" + deploymentName + "' not deleted, will retry");
+        } else {
+            LOG.info("Deployment '" + deploymentName + "' deleted!");
```
*FlinkDeployment

gyfora reviewed (changes requested):
Added some logging related comments.

We should also handle the `stateless` case a bit more gracefully. A simple stateless example fails with errors like:

```
flink-kubernetes-operator-7f95587856-cz8j7 2025-07-18 06:24:24,708 i.j.o.p.e.EventProcessor       [WARN ][default/basic-example] Uncaught error during event processing ExecutionScope{ resource id: ResourceID{name='basic-example', namespace='default'}, version: 11407039} - but another reconciliation will be attempted because a superseding event has been received or another retry attempt is pending.
flink-kubernetes-operator-7f95587856-cz8j7 org.apache.flink.kubernetes.operator.exception.UpgradeFailureException: Latest checkpoint not externally addressable, manual recovery required.
flink-kubernetes-operator-7f95587856-cz8j7     at org.apache.flink.kubernetes.operator.service.AbstractFlinkService.getLastCheckpoint(AbstractFlinkService.java:569)
flink-kubernetes-operator-7f95587856-cz8j7     at org.apache.flink.kubernetes.operator.controller.FlinkBlueGreenDeploymentController.configureSavepoint(FlinkBlueGreenDeploymentController.java:415)
flink-kubernetes-operator-7f95587856-cz8j7     at org.apache.flink.kubernetes.operator.controller.FlinkBlueGreenDeploymentController.checkAndInitiateDeployment(FlinkBlueGreenDeploymentController.java:378)
flink-kubernetes-operator-7f95587856-cz8j7     at org.apache.flink.kubernetes.operator.controller.FlinkBlueGreenDeploymentController.reconcile(FlinkBlueGreenDeploymentController.java:117)
flink-kubernetes-operator-7f95587856-cz8j7     at org.apache.flink.kubernetes.operator.controller.FlinkBlueGreenDeploymentController.reconcile(FlinkBlueGreenDeploymentController.java:65)
flink-kubernetes-operator-7f95587856-cz8j7     at io.javaoperatorsdk.operator.processing.Controller$1.execute(Controller.java:161)
flink-kubernetes-operator-7f95587856-cz8j7     at io.javaoperatorsdk.operator.processing.Controller$1.execute(Controller.java:117)
flink-kubernetes-operator-7f95587856-cz8j7     at org.apache.flink.kubernetes.operator.metrics.OperatorJosdkMetrics.timeControllerExecution(OperatorJosdkMetrics.java:82)
flink-kubernetes-operator-7f95587856-cz8j7     at io.javaoperatorsdk.operator.processing.Controller.reconcile(Controller.java:116)
flink-kubernetes-operator-7f95587856-cz8j7     at io.javaoperatorsdk.operator.processing.event.ReconciliationDispatcher.reconcileExecution(ReconciliationDispatcher.java:153)
flink-kubernetes-operator-7f95587856-cz8j7     at io.javaoperatorsdk.operator.processing.event.ReconciliationDispatcher.handleReconcile(ReconciliationDispatcher.java:130)
flink-kubernetes-operator-7f95587856-cz8j7     at io.javaoperatorsdk.operator.processing.event.ReconciliationDispatcher.handleDispatch(ReconciliationDispatcher.java:97)
flink-kubernetes-operator-7f95587856-cz8j7     at io.javaoperatorsdk.operator.processing.event.ReconciliationDispatcher.handleExecution(ReconciliationDispatcher.java:68)
flink-kubernetes-operator-7f95587856-cz8j7     at io.javaoperatorsdk.operator.processing.event.EventProcessor$ReconcilerExecutor.run(EventProcessor.java:475)
flink-kubernetes-operator-7f95587856-cz8j7     at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
flink-kubernetes-operator-7f95587856-cz8j7     at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
flink-kubernetes-operator-7f95587856-cz8j7     at java.base/java.lang.Thread.run(Unknown Source)
```

during upgrades.

In general if an upgrade would go through successfully for a regular FlinkDeployment it should also go through for a BlueGreenDeployment.

schongloo commented:
> Added some logging related comments.
> 
> We should also handle the `stateless` case a bit more gracefully. A simple stateless example fails with errors like:
> 
> ```
> ... org.apache.flink.kubernetes.operator.exception.UpgradeFailureException: Latest checkpoint not externally addressable, manual recovery required.
> ```
> 
> during upgrades.
> 
> In general if an upgrade would go through successfully for a regular FlinkDeployment it should also go through for a BlueGreenDeployment.

While not necessarily related to the `stateless` case. In this situation it's an error I throw when checkpoints are not externally accessible (which is mandatory), the confusing part I throw it during the upgrade only. I should also validate this type of restriction(s) before the initial deployment.

I've addressed all the simple refactoring/logging comments, remaining to be done asap:

schongloo commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkBlueGreenDeploymentController.java:
```diff
@@ -0,0 +1,595 @@
+        // Deployment
+        FlinkDeployment flinkDeployment = new FlinkDeployment();
+        flinkDeployment.setApiVersion("flink.apache.org/v1beta1");
+        flinkDeployment.setKind("FlinkDeployment");
```
No, remains of early tests, removed.

gyfora commented:
> > Added some logging related comments.
> > We should also handle the `stateless` case a bit more gracefully. A simple stateless example fails with errors like:
> > ```
> > ... org.apache.flink.kubernetes.operator.exception.UpgradeFailureException: Latest checkpoint not externally addressable, manual recovery required.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > during upgrades.
> > In general if an upgrade would go through successfully for a regular FlinkDeployment it should also go through for a BlueGreenDeployment.
> 
> While not necessarily related to the `stateless` case. In this situation it's an error I throw when checkpoints are not externally accessible (which is mandatory), the confusing part I throw it during the upgrade only. I should also validate this type of restriction(s) before the initial deployment.
> 
> I've addressed all the simple refactoring/logging comments, remaining to be worked on asap: diff builder and e2e test

@schongloo , `stateless` Blue/Green upgrades should not use any state (even if available) and simply start the new job from an empty state. It should definitely not throw any error if checkpointing/external checkpoints are disabled

davidradl commented on docs/content/docs/custom-resource/reference.md:
```diff
@@ -57,6 +57,23 @@ This serves as a full reference for FlinkDeployment and FlinkSessionJob custom r
 | Parameter | Type | Docs |
 | ----------| ---- | ---- |
 
+### FlinkBlueGreenDeploymentConfigOptions
```
I think we should define the Blue Green deployment concept some where in the docs, ideally with diagrams.

davidradl commented on flink-kubernetes-operator-api/src/main/java/org/apache/flink/kubernetes/operator/api/bluegreen/DeploymentType.java:
```diff
@@ -0,0 +1,39 @@
+ * Enumeration of the two possible Flink Blue/Green deployment types. Only one of each type will be
+ * present at all times for a particular job.
+ */
+public enum DeploymentType {
```
I suggest the enum has blue green in it's name

gyfora commented on docs/content/docs/custom-resource/reference.md:
```diff
@@ -78,6 +95,17 @@ This serves as a full reference for FlinkDeployment and FlinkSessionJob custom r
 | logConfiguration | java.util.Map<java.lang.String,java.lang.String> | Log configuration overrides for the Flink deployment. Format logConfigFileName -> configContent. |
 | mode | org.apache.flink.kubernetes.operator.api.spec.KubernetesDeploymentMode | Deployment mode of the Flink cluster, native or standalone. |
 
+### FlinkDeploymentTemplateSpec
```
I think that makes sense, because this is actually a template for a FlinkDeployment. 
In the same way the PodTemplateSpec doesn't have any reference to the actual CR-s that use it such as Deployment/ReplicaSet

davidradl commented on flink-kubernetes-operator-api/src/main/java/org/apache/flink/kubernetes/operator/api/spec/FlinkBlueGreenDeploymentConfigOptions.java:
```diff
@@ -0,0 +1,73 @@
+     *   <li>MINUTES： "m", "min", "minute"
+     *   <li>SECONDS： "s", "sec", "second"
+     *   <li>MILLISECONDS： "ms", "milli", "millisecond"
+     *   <li>MICROSECONDS： "µs", "micro", "microsecond"
```
I am concerned about `µs` as it is a non standard character. [Flink docs](https://nightlies.apache.org/flink/flink-docs-release-1.20/docs/deployment/config/) use `us`. I suggest we use that here to be consistent.

davidradl commented on flink-kubernetes-operator-api/src/main/java/org/apache/flink/kubernetes/operator/api/spec/FlinkBlueGreenDeploymentConfigOptions.java:
```diff
@@ -0,0 +1,73 @@
+                    .durationType()
+                    .defaultValue(Duration.ofMinutes(10))
+                    .withDescription(
+                            "The max time to wait in milliseconds for a deployment to become ready before aborting it. Cannot be smaller than 10 minutes.");
```
nit: it is a duration type not milliseconds

davidradl commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkBlueGreenDeploymentController.java:
```diff
@@ -0,0 +1,142 @@
+ * <p>Orchestration Process
+ *
+ * <p>FlinkBlueGreenDeploymentController.reconcile() ↓ 1. Create BlueGreenContext with current
+ * deployment state ↓ 2. Query StateHandlerRegistry for appropriate handler ↓ 3. Delegate to
```
what are the `↓` characters?

schongloo commented:
@davidradl thank you for all your feedback. The past few weeks I was focused on mostly the high level refactoring, I'll address your comments ASAP.

vikasap commented on flink-kubernetes-operator-api/src/main/java/org/apache/flink/kubernetes/operator/api/spec/FlinkBlueGreenDeploymentConfigOptions.java:
```diff
@@ -0,0 +1,73 @@
+                    .withDescription(
+                            "Configurable delay to use when the operator reschedules a reconciliation.");
+
+    public static final ConfigOption<Duration> DEPLOYMENT_DELETION_DELAY =
```
These configs are not being used. Is there an intent to use them later ?

schongloo replied:
Not sure I follow. These values are used throughout the `BlueGreenDeploymentService`.

vikasap commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/bluegreen/BlueGreenDeploymentService.java:
```diff
@@ -0,0 +1,659 @@
+            return getSavepointObject(ctx, savepointFetchResult.getLocation());
+        }
+
+        // Currently not using last checkpoint recovery for LAST_STATE upgrade mode
```
Any particular reason this is disabled explicitly ?

schongloo replied:
We realized using checkpoints to preserve state across Blue/Green transitions can be a brittle strategy since it would depend entirely on the users configuration (whether they're even enabled or not, externally retained, for how long, etc.). So we opted to go for Savepoints instead, at least initially.

maheepm-lyft commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/bluegreen/BlueGreenDeploymentService.java:
```diff
@@ -0,0 +1,659 @@
+    private UpdateControl<FlinkBlueGreenDeployment> patchFlinkDeployment(
+            BlueGreenContext context, BlueGreenDeploymentType blueGreenDeploymentTypeToPatch) {
+
+        String childDeploymentName =
```
Hi, thanks for following up! Yes I'm aware of the PR, the committer is in my org and we collaborated on that change. Thanks again for pushing out this blue green feature.
